backends/p4tools/modules/testgen/core/small_step/table_stepper.cpp

```diff
@@ -184,6 +184,7 @@ const IR::Expression *TableStepper::evalTableConstEntries() {
         const auto *actionType = stepper->state.getP4Action(tableAction);
         auto &nextState = stepper->state.clone();
         nextState.markVisited(entry);
+        nextState.markVisited(actionType);
```

Comment:
Hmm, this is actually a very specific type of coverage. This does not cover when an action is called directly and not via a table, for example. Is this intended? 

Otherwise, I would just move the markVisited call into the P4Action visitor.

Reply:
Hmm. I'm not sure how the testgen's midend is configured and if there will be directly-called actions left, I should check it (I presumed there will be none). If there can be actions called directly, it would make sense to handle those too probably (presumably they would be still included in statements coverage because of the call, but it would still make sense to have them in actions coverage too).

A slight problem is that I actually have to mark the action as visited before the step into it is generate as in many places the table stepper creates a new, cloned action out of the original, with the arguments given by the control plane config; https://github.com/p4lang/p4c/blob/47862844a1599e0ca3e8cb31c84be894c083d48f/backends/p4tools/modules/testgen/core/small_step/table_stepper.cpp#L346-L348 In these cases marking the action as visited only when the call is interpreted would yield wrong results I believe.

Reply:
OK, `MoveActionsToTables` is not running in testgen, let me think about a way to handle directly called actions too.

Reply:
>A slight problem is that I actually have to mark the action as visited before the step into it is generate as in many places the table stepper creates a new, cloned action out of the original, with the arguments given by the control plane config; 

That should not be a problem for the markVisited function since we are comparing entries based on the valid source information? If we mismatch or add duplicate entries that looks like a bug.

Reply:
I see. I did not realize we use the source info for identity. In that case this is actually quite simple.